acir/acir.cpp

```diff
@@ -0,0 +1,5260 @@
+    };
+
+    struct Opcode {
+
+        struct Arithmetic {
+            Circuit::Expression value;
+
+            friend bool operator==(const Arithmetic&, const Arithmetic&);
+            std::vector<uint8_t> bincodeSerialize() const;
+            static Arithmetic bincodeDeserialize(std::vector<uint8_t>);
+        };
+
+        struct BlackBoxFuncCall {
+            Circuit::BlackBoxFuncCall value;
+
+            friend bool operator==(const BlackBoxFuncCall&, const BlackBoxFuncCall&);
+            std::vector<uint8_t> bincodeSerialize() const;
+            static BlackBoxFuncCall bincodeDeserialize(std::vector<uint8_t>);
+        };
+
+        struct Directive {
+            Circuit::Directive value;
+
+            friend bool operator==(const Directive&, const Directive&);
+            std::vector<uint8_t> bincodeSerialize() const;
+            static Directive bincodeDeserialize(std::vector<uint8_t>);
+        };
+
+        struct Block {
+            Circuit::MemoryBlock value;
+
+            friend bool operator==(const Block&, const Block&);
+            std::vector<uint8_t> bincodeSerialize() const;
+            static Block bincodeDeserialize(std::vector<uint8_t>);
+        };
+
+        struct ROM {
+            Circuit::MemoryBlock value;
+
+            friend bool operator==(const ROM&, const ROM&);
+            std::vector<uint8_t> bincodeSerialize() const;
+            static ROM bincodeDeserialize(std::vector<uint8_t>);
+        };
+
+        struct RAM {
+            Circuit::MemoryBlock value;
+
+            friend bool operator==(const RAM&, const RAM&);
+            std::vector<uint8_t> bincodeSerialize() const;
+            static RAM bincodeDeserialize(std::vector<uint8_t>);
+        };
+
+        struct Brillig {
+            Circuit::Brillig value;
+
+            friend bool operator==(const Brillig&, const Brillig&);
+            std::vector<uint8_t> bincodeSerialize() const;
+            static Brillig bincodeDeserialize(std::vector<uint8_t>);
+        };
+
+        std::variant<Arithmetic, BlackBoxFuncCall, Directive, Block, ROM, RAM, Brillig> value;
+
+        friend bool operator==(const Opcode&, const Opcode&);
+        std::vector<uint8_t> bincodeSerialize() const;
+        static Opcode bincodeDeserialize(std::vector<uint8_t>);
+    };
+
+    struct PublicInputs {
+        std::vector<Circuit::Witness> value;
+
+        friend bool operator==(const PublicInputs&, const PublicInputs&);
+        std::vector<uint8_t> bincodeSerialize() const;
+        static PublicInputs bincodeDeserialize(std::vector<uint8_t>);
+    };
+
+    struct Circuit {
+        uint32_t current_witness_index;
+        std::vector<Circuit::Opcode> opcodes;
+        Circuit::PublicInputs public_parameters;
+        Circuit::PublicInputs return_values;
+
+        friend bool operator==(const Circuit&, const Circuit&);
+        std::vector<uint8_t> bincodeSerialize() const;
+        static Circuit bincodeDeserialize(std::vector<uint8_t>);
```

Comment:
bberg would then just have to iterate through the opcodes field and construct a `ConstraintSystem` from this.